src/shared/Core/Trace2.cs

```diff
@@ -0,0 +1,93 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+using Newtonsoft.Json.Serialization;
+
+namespace GitCredentialManager;
+
+// The different event types tracked in the TRACE2 tracing
+// system.
+public enum Event {}
+
+/// <summary>
+/// Represents the application's TRACE2 tracing system, which
+/// writes to a collector service that stores the trace information.
+/// </summary>
+public interface ITrace2 : IDisposable
+{
+    /// <summary>
+    /// Add a listener to the trace writer.
+    /// </summary>
+    /// <param name="listener">The listener to add.</param>
+    void AddListener(CollectorWriter listener);
+}
+
+public class Trace2: DisposableObject, ITrace2
+{
+    private const string ParentSidVariable = "GIT_TRACE2_PARENT_SID";
+    private const string CurrentSidVariable = "GCM_TRACE2_PARENT_SID";
+    private readonly ICommandContext _context;
+    private bool _isDisposed;
+    private CollectorWriter _writer;
+
+    public void AddListener(CollectorWriter listener)
+    {
+        ThrowIfDisposed();
+        
+        // Try not to add the same listener more than once
+        if (_writer == listener)
+            return;
+
+        _writer = listener;
+        _writer.Start();
+    }
+
+    public new void Dispose()
+    {
+        if (_isDisposed)
+        {
+            // This instance has already been disposed
+            return;
+        }
+        
+        _writer.Dispose();
+        _isDisposed = true;
+    }
+
+    private void WriteMessage(CollectorMessage message)
+    {
+        ThrowIfDisposed();
+        
+        if (_writer == null)
+        {
+            return;
+        }
+        
+        // Return rather than throwing an exception by
+        // attempting to write to a failed connection.
+        if (_writer.State == CollectorWriterState.Failing)
+        {
+            return;
+        }
+
+        try
+        {
+            _writer?.TryEnqueue(message.ToJson());
+        }
+        catch { /* squelch */ }
+    }
+}
+
+public abstract class CollectorMessage
+{
+    [JsonProperty("event")] public Event Event { get; set; }
+    [JsonProperty("sid")] public string Sid { get; set; }
+    [JsonProperty("thread")] public string Thread { get; set; }
+    [JsonProperty("time")] public string Time { get; set; }
+    [JsonProperty("file")] public string File { get; set; }
+    [JsonProperty("line")] public double Line { get; set; }
```

Comment:
I know many default IDE formatting rules put these attributes one the same line as the property, but  think it's better to have them in the more traditional style:


```csharp
[JsonProperty("event")]
public Event Event { get; set; }
```

Reply:
Sure, that's an easy change, and I like the look of it better as well.

Reply:
Easier to stack additional attributes if we need to as well.